### PR TITLE
AB#4077 added the code/endpoint to create a user on the api project

### PR DESCRIPTION
**Pull request by ESDC-Olivier-Roy**
### Description
added the code/endpoint to create a user on the api project.
The endpoint is exposed and now we can add a user with its raoidc user Id and the email address with it.

### Related Azure Boards Work Items
[AB#4077](https://dev.azure.com/DTS-STN/1fc40a8f-28cf-47bc-b6e4-1c234bd06177/_workitems/edit/4077)

### Checklist

- [x] I have tested the changes locally
- [x] I have updated the documentation if necessary
- [x] I have added/updated tests that prove my fix is effective or that my feature works
- [ ] I have checked that my code follows the project's coding style by running `npm run format:check`
- [ ] I have checked that my code contains no linting errors by running `npm run lint`
- [ ] I have checked that my code contains no type errors by running `npm run typecheck`
- [ ] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [ ] I have checked that all e2e tests pass by running `npm run test:e2e`

### CODE REVIEW

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserCreateModel.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+@Immutable
+@Schema(name = "UserCreate")
+@JsonDeserialize(as = ImmutableUserCreateModel.class)
+public interface UserCreateModel {
+	
+	@Email
+	String getEmail();
+
+	List<UserAttributeModel> getUserAttributeModel();
```
Comment (nicholas-ly) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserCreateModel.java:17]: We should create a new, immutable, model for creating a user attribute, which consists of validation annotations to ensure that the name/value pair are not blank.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 2]
```diff
@@ -197,6 +205,19 @@ public boolean verifyEmail(String userId, String code){
 		return true;
 	}
 
+	public User createUser(String userEmail, List<UserAttributeModel> userAttributesModel){
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:208]: I think this should be `public User createUser(User user)`. Also, service-layer classes should not know anything about web-layer models.

All service methods should accept only domain objects.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 3]
```diff
@@ -197,6 +205,19 @@ public boolean verifyEmail(String userId, String code){
...
+		//Verify Email
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:209]: This comment should be removed.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 4]
```diff
@@ -197,6 +205,19 @@ public boolean verifyEmail(String userId, String code){
...
+		Assert.hasText(userEmail, "userEmail is required; it must not be null or blank");
+
+		Collection<UserAttributeEntity> userAttributesCollection = new ArrayList<UserAttributeEntity>();
+		
+		for(UserAttributeModel e : userAttributesModel){
+			userAttributesCollection.add(new UserAttributeEntityBuilder().name(e.getName()).value(e.getValue()).build());
+		}
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:216]: This code should be removed once the method is updated to accept a `User` instance.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 5]
```diff
@@ -197,6 +205,19 @@ public boolean verifyEmail(String userId, String code){
...
+		return userMapper.toUser(userRepository.save(new UserEntityBuilder().email(userEmail)
+			.userAttributes(userAttributesCollection).build()));
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:218]: Use a mapper method to convert from `User` to `UserEntity`. Only in rare cases should you ever have to use `UserEntityBuilder`.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 6]
```diff
@@ -64,4 +69,13 @@ public void updateUserById(
 		userService.updateUser(user.getId(), userUpdateModel.getEmail());
 	}	
 
+	@PostMapping
+	@Operation(summary = "Create a new user ")
+	public void createUserByEmailAndUserAttributes(
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:74]: This method should have a `@ResponseStatus(HttpStatus.NO_CONTENT)` to indicate that it returns a 204 status code.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 7]
```diff
@@ -64,4 +69,13 @@ public void updateUserById(
 		userService.updateUser(user.getId(), userUpdateModel.getEmail());
 	}	
 
+	@PostMapping
+	@Operation(summary = "Create a new user ")
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:73]:
```suggestion
	@Operation(summary = "Create a new user")
```

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 8]
```diff
@@ -64,4 +69,13 @@ public void updateUserById(
...
+	public void createUserByEmailAndUserAttributes(
+
+		@Validated @RequestBody UserCreateModel user) {
+			List<UserAttributeModel> userAttributesModels = user.getUserAttributeModel();
+
+			userService.createUser(user.getEmail(), userAttributesModels);
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:79]: You should use a mapper method to convert from the model to the domain object, then call `userService.createUser(..)` with that domain object.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserCreateModel.java (hunk 1 above)
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserCreateModel.java:17]: The convention is to call this `getUserAttributes()`.

**Review by gregory-j-baker — DISMISSED** [review]
Good PR.. nice and small. :chefs-kiss:

**Comment by ESDC-Olivier-Roy** [comment]
@nicholas-ly @gregory-j-baker heyhey, i've added some changes to the code, can you review it and see if that resolves the comments this PR has so far?

Thanks

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 9]
```diff
@@ -20,6 +23,8 @@
 import io.swagger.v3.oas.annotations.responses.ApiResponse;
 import io.swagger.v3.oas.annotations.tags.Tag;
 import jakarta.validation.constraints.NotBlank;
+import org.springframework.web.bind.annotation.PostMapping;
+import ca.gov.dtsstn.cdcp.api.web.v1.model.UserAttributeModel;
```
[the remark this answers is not in the document]
Comment (ESDC-Olivier-Roy) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java]: ive added this settings to my vsCode

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 10]
```diff
@@ -86,4 +92,15 @@ public void updateUserById(
...
+	@PostMapping
+	@ResponseStatus(HttpStatus.NO_CONTENT)
+	@Operation(summary = "Create a new user")
+	public void createUserByEmailAndUserAttributes(@Validated @RequestBody UserCreateModel userCreateModel) 
+	{
+		UserModel userModel = new UserModel();
+		userModel.setUserAttributes(userCreateModel.getUserAttributes());
+		userModel.setEmail(userCreateModel.getEmail());
+		
+		userService.createUser(userModelMapper.toDomainObject(userModel));
```
Comment (nicholas-ly) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:104]: The mapper should map from `UserCreateModel` (not `UserModel)` to `User`. That way, this method won't have to map from `UserCreateModel` --> `UserModel` --> `User`

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 11]
```diff
@@ -33,6 +38,7 @@
 public class UsersController {
 
 	private final UserModelMapper userModelMapper = Mappers.getMapper(UserModelMapper.class);
+	private final UserAttributeMapper userAttributeMapper = Mappers.getMapper(UserAttributeMapper.class);
```
Comment (nicholas-ly) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:41]: `UserAttributeMapper` is being referenced but it is not being used. Can we remove this line?

Comment (ESDC-Olivier-Roy) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java]: removed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java (hunk 10 above)
Comment (ESDC-Olivier-Roy) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java]: new model created

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserCreateModel.java (hunk 1 above)
Comment (nicholas-ly) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserCreateModel.java]: Although `UserAttributeCreateModel` has been created, it is currently mutable (setter methods exist and fields not marked as `final`)

**Review by nicholas-ly — APPROVED** [review]

**Pull request merged**